### PR TITLE
MinSR algorithm

### PR DESCRIPTION
**Pull request by chrisrothUT**
This is a draft of the MinSR algorithm relegated to `netket.experimental` for the time being. 

The MinSR algorithm involves contracting the Jacobian with itself over the parameter dimension to compute the neural tangent kernel (NTK). There is a memory cost N(Y + P^l) where N is the number of samples, Y is the size of the total hidden state across all layers, and P^l is the parameters per layer. This can be reduced by computing the NTK matrix in chunks using the argument `jacobian_chunk_size`.

### CODE REVIEW

**Comment by gcarleo** [comment]
hi Chris what's the status of this? did you manage to make it work?

**Comment by yannra** [comment]
Hi all,
Yes - we have implemented a quick (and not at all documented) test of the minSR updates just to see how well it works. Our implementation should work but it's definitely still a bit rough around the edges (in particular for non-holomorphic, complex models the maths should be checked).

Also beyond the minSR implementation, we always wanted to reach out to see if and what of our features might be useful enough to merge into the NetKet core. Our GPSKet repository is mostly just a playground for us to test different ideas (which currently is not well documented and certainly does not meet the high test standard/coding practices of the NetKet codebase). We could/should maybe initiate that discussion somewhere else.

Re. the minSR implementation: We are obviously happy to contribute to this PR (which does however seem to follow a slightly different design than our admittedly slightly hacky implementation). Let me/us know if there is any info you need about our code or if there is any other way we can contribute.

Best,
Yannic

**Comment by chrisrothUT** [comment]
Okay, I think the code should work generally now. A few points

1) I used a separate parameter for `chunk_size` and `grad_chunk_size`. These typically have different memory requirements as grad requires storing the hidden state of all of the layers. 

2) When `grad_chunk_size` < `n_samples` the NTK can't be computed all at once and fit into memory. In this case the NTK matrix is computed in blocks, which of course duplicates some computation. There are ways to compute the gradient by storing one hidden layer at a time that makes it less memory intensive, but I've found this frustrating to get working. Perhaps @ChenAo-Phys has some ideas. 

3) I don't think this can be integrated with the preconditioner interface. This is because the NTK acts in the sample subspace, not the parameter subspace. Maybe this should be a different type like `sample_preconditioner`?

4) I still have not implemented MPI. Perhaps someone else can do this in a different PR?

**Comment by gcarleo** [comment]
thank you Chris, tests are missing, can you add some so to make sure this works as expected? @yannra does your version feature MPI ? Looks like also the London version is much more compact in terms of mere lines of code, I wonder if this PR cannot be made less verbose maybe? (for example, do we really need a class for the "NTK" in this case?)

File: netket/experimental/vqs/MinSR.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+    model_state: PyTree,
+    σ: jnp.ndarray,
+    grad_mean: jnp.ndarray,
+) -> Tuple[PyTree, PyTree]:
+
+    n_samples = len(σ)
+
+    NTK = jnp.zeros([n_samples, n_samples], dtype="complex")
+
+    for i in range(n_samples // jcs):
```
Comment (gcarleo) [inline netket/experimental/vqs/MinSR.py:165]: isn't this double for loop very inefficient?

**Comment by gcarleo** [comment]
pinging @inailuig , the master of the NTK

File: netket/experimental/vqs/MinSR.py (hunk 1 above)
Comment (chrisrothUT) [reply netket/experimental/vqs/MinSR.py:165]: Yes it's most efficient to compute the NTK all in one go but you need to chunk the matrix into parts if the gradient function takes too much memory.

**Comment by chrisrothUT** [comment]
> Looks like also the London version is much more compact in terms of mere lines of code, I wonder if this PR cannot be made less verbose maybe? (for example, do we really need a class for the "NTK" in this case?)

The two versions are pretty functionally different. The GPSKet version stores the Jacobian in memory and solves a linear system to find OO^Tε,  whereas my version computes the NTK and stores it in memory. The NTK requires a factor of `n_samples/n_parameters` less memory. 

I also used a direct pseudoinverse instead of solving a linear system with conjugate gradient, this avoids the need for a diagonal shift.

**Comment by chrisrothUT** [comment]
I'm quite busy writing my thesis and haven't used MPI in a while, does anyone want to work with me on this?

**Comment by yannra** [comment]
> @yannra does your version feature MPI ?

Yes - our version parallelizes the evaluation of local energies and the computation of the derivatives over the MPI ranks. However, then a global Jacobian for all samples and parameters is constructed via an allreduce and the update equation is solved on each rank at the same time (which could be improved but should ultimately not be a massive problem as solving the update equation would typically be quite fast for moderate numbers of samples).

>  The GPSKet version stores the Jacobian in memory and solves a linear system to find OO^Tε, whereas my version computes the NTK and stores it in memory. The NTK requires a factor of n_samples/n_parameters less memory.

This issue pointed out by @chrisrothUT, on the other hand, is definitely valid and you will run into memory limitations faster. With n_samples*n_parameters < ~10^8 that should all be fine though.

> I also used a direct pseudoinverse instead of solving a linear system with conjugate gradient, this avoids the need for a diagonal shift.

Well, I mean you can pass whatever solver you want in our code, to be honest there is no particular reason why I put the jnp.linalg.lstsq as the default rather than the pseudoinverse. The diagonal shift is completely optional just to have more regularisation options to play around with (in our, not at all comprehensive, tests it seemed to be more stable in the limit of few parameters than applying the truncation in the singular values when constructing a pseudo-inverse).

Anyway, let me see if I can maybe add MPI support to this PR to have that (experimentally) supported (and we can retire our hacky test implementation). That will definitely take me a couple of days though as I won't have time till next week.

**Comment by PhilipVinc** [comment]
@yannra @chrisrothUT I would be super happy to merge this PR if you can get MPI working (it should just be about adding a few calls here and there) and, especially, if you add a few tests.

If you add this things I will quickly review and merge!

-- 

I also have a question: how should we chunk? 
The easiest way to implement minSR/NTK is to use `nk.jax.jacobian` as you did in GPSKet. 
This approach would be chunking the calculation of the jacobian, but you still need the memory to store the whole jacobian and NTK matrix at once which means we have at least the memory requirement of

$$
N_s * N_p + N_s^2
$$

In this implementation by @chrisrothUT ,the calculation of the NTK is chunked, meaning that  each entry of the Jacobian is recomputed `n_chunk` times.
My question for Chris is: is this actually necessary, or are we being overly careful? 
It's quite more expensive, so I want to make sure it makes sense...

**Comment by chrisrothUT** [comment]
I think you've hit on a key point here. When the computation of grad doesn't fit into memory there's a tradeoff between memory and compute time. My algorithm requires an extra `n_chunks/2` calls to grad whereas Yannic's requires an extra Ns*Np memory. 

I think both versions are pretty useful, so we might just want to have both. Maybe I'll finish this PR and then Yannic can add his?

> My question for Chris is: is this actually necessary, or are we being overly careful?

I don't think so. If you want to train big models the Ns*Np bound is prohibitive. I've been able to go from 10^5 to 10^7 parameters using the MinSr algorithm

> if you can get MPI working

I think it's a bit trickier to use MPI with my algorithm. The best way is probably to compute part of the NTK on each device `[n_samples/mpi.n_nodes, n_samples]`? I'll try to work on it in the next few days

**Comment by PhilipVinc** [comment]
Isn't your approach calling jacobian $\frac{n_\text{chunks}*(n_\text{chunks}+1)}{2}$ vs the standard approach that would call jacobian only `n_chunks` times?

**Comment by gcarleo** [comment]
@chrisrothUT please have a look at #1623 , that one implements also MPI and should be more general, so unless you see issues we would close this one in favor of the latest PR

**Comment by PhilipVinc** [comment]
Superseded by #1623 .
